OrbitLinuxTracing/UprobesUnwindingVisitor.cpp

```diff
@@ -135,6 +135,28 @@ void UprobesUnwindingVisitor::visit(StackSamplePerfEvent* event) {
 }
 
 void UprobesUnwindingVisitor::visit(UprobesWithStackPerfEvent* event) {
+  // We are seeing that on thread migration, uprobe events can sometimes be
+  // duplicated. The idea of the workaround is that for a given thread's
+  // sequence of u(ret)probe events, two consecutive uprobe events must be
+  // associated with decreasing stack pointers (nested function calls, stack
+  // grows by decreasing stack pointer).  If an extra uprobe event is generated,
+  // then the second uprobe event will be associated with a stack pointer that
+  // is greater or equal to the previous stack pointer, and that's not normal.
+  // In that situation, we discard the second uprobe event.
+
+  // Duplicate uprobe detection.
+  uint64_t uprobe_sp = event->GetRegisters()[PERF_REG_X86_SP];
+  std::vector<uint64_t>& uprobe_sps = uprobe_sp_per_thread_[event->GetTid()];
+  if (!uprobe_sps.empty()) {
+    uint64_t last_uprobe_sp = uprobe_sps.back();
+    uprobe_sps.pop_back();
+    if (uprobe_sp >= last_uprobe_sp) {
```

Comment:
What if we also checked that the function is the same? Shouldn't that cover more cases with functions that don't use the stack?

Reply:
Yes, I think that's true.

Reply:
I'll submit this change as is (with the variable naming change) since it's already a good improvement and add the suggested test.  The plan is to also look at the instruction pointer, it should be the same for both events.  That being said, I still see the issue happening sometimes, so we also have to figure out why.